File: domain/src/test/java/de/seuhd/campuscoffee/domain/implementation/ReviewServiceTest.java
package de.seuhd.campuscoffee.domain.implementation;

import de.seuhd.campuscoffee.domain.configuration.ApprovalConfiguration;
import de.seuhd.campuscoffee.domain.exceptions.NotFoundException;
import de.seuhd.campuscoffee.domain.exceptions.ValidationException;
import de.seuhd.campuscoffee.domain.model.objects.Pos;
import de.seuhd.campuscoffee.domain.model.objects.Review;
import de.seuhd.campuscoffee.domain.model.objects.User;
import de.seuhd.campuscoffee.domain.ports.data.PosDataService;
import de.seuhd.campuscoffee.domain.ports.data.ReviewDataService;
import de.seuhd.campuscoffee.domain.ports.data.UserDataService;
import de.seuhd.campuscoffee.domain.tests.TestFixtures;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Objects;

import static de.seuhd.campuscoffee.domain.tests.TestFixtures.getApprovalConfiguration;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

/**
 * Unit and integration tests for the operations related to reviews.
 */
@ExtendWith(MockitoExtension.class)
public class ReviewServiceTest {
    private final ApprovalConfiguration approvalConfiguration = getApprovalConfiguration();

    @Mock
    private ReviewDataService reviewDataService;

    @Mock
    private UserDataService userDataService;

    @Mock
    private PosDataService posDataService;

    private ReviewServiceImpl reviewService;

    @BeforeEach
    void beforeEach() {
        reviewService = new ReviewServiceImpl(
                reviewDataService, userDataService, posDataService, approvalConfiguration
        );
    }

    /**
     * Verifies a user cannot approve their own review. The test verifies that when the author
     * attempts to approve their own review, an exception is thrown.
     */
    @Test
    void approvalFailsIfUserIsAuthor() {
        // given
        Review review = TestFixtures.getReviewFixtures().getFirst();
        assertNotNull(review.author().id());
        when(userDataService.getById(review.author().id())).thenReturn(review.author());
        assertNotNull(review.id());
        when(reviewDataService.getById(review.id())).thenReturn(review);

        // when, then
        assertThrows(ValidationException.class, () -> reviewService.approve(review.getId(), review.author().getId()));
        verify(userDataService).getById(review.author().id());
        verify(reviewDataService).getById(review.getId());
    }

    /**
     * Verifies that the approval succeeds if the approver is not the author.
     * The test verifies that a valid approval increments the approval count
     * and sets the status of the review to "approved" if the required threshold is met.
     */
    @Test
    void approvalSuccessfulIfUserIsNotAuthor() {
        // given
        // one short of the quorum, so the single approval below pushes it to exactly the quorum
        Review review = TestFixtures.getReviewFixtures().getFirst().toBuilder()
                .approvalCount(approvalConfiguration.minCount() - 1)
                .approved(false)
                .build();
        User user = TestFixtures.getUserFixtures().getLast();
        assertNotNull(user.getId());
        when(userDataService.getById(user.getId())).thenReturn(user);
        assertNotNull(review.getId());
        when(reviewDataService.getById(review.getId())).thenReturn(review);
        when(reviewDataService.upsert(any(Review.class))).thenAnswer(
                invocation -> invocation.getArgument(0)
        );

        // when
        Review approvedReview = reviewService.approve(review.getId(), user.getId());

        // then
        verify(userDataService).getById(user.getId());
        verify(reviewDataService).getById(review.getId());
        verify(reviewDataService).upsert(any(Review.class));
        assertThat(approvedReview.approvalCount()).isEqualTo(review.approvalCount() + 1);
        assertThat(approvedReview.approved()).isTrue();
    }

    /**
     * Verifies that the service returns all approved reviews for a given POS.
     */
    @Test
    void retrieveAllApprovedPos() {
        // given
        Pos pos = TestFixtures.getPosFixtures().getFirst();
        assertNotNull(pos.getId());
        List<Review> reviews = TestFixtures.getReviewFixtures().stream()
                .map(review -> review.toBuilder()
                        .pos(pos)
                        .approvalCount(approvalConfiguration.minCount())
                        .approved(true)
                        .build())
                .toList();
        when(posDataService.getById(pos.getId())).thenReturn(pos);
        when(reviewDataService.filter(pos, true)).thenReturn(reviews);

        // when
        List<Review> retrievedReviews = reviewService.filter(Objects.requireNonNull(pos.getId()), true);

        // then
        verify(posDataService).getById(pos.getId());
        verify(reviewDataService).filter(pos, true);
        assertThat(retrievedReviews).hasSize(reviews.size());
    }

    /**
     * Verifies that an exception is thrown when attempting to create or update a review
     * for a POS that does not exist in the system.
     */
    @Test
    void createReviewPosDoesNotExistException() {
        // given
        Review review = TestFixtures.getReviewFixtures().getFirst();
        assertNotNull(review.pos().getId());
        when(posDataService.getById(review.pos().getId())).thenThrow(
                new NotFoundException(review.pos().getClass(), review.pos().getId())
        );

        // when, then
        assertThrows(NotFoundException.class, () -> reviewService.upsert(review));
        verify(posDataService).getById(review.pos().getId());
    }

    /**
     * Verifies that a user cannot create more than one review for the same POS.
     * If an existing review for the same POS by the same author is found, the upsert should fail
     * with a validation error.
     */
    @Test
    void userCannotCreateMoreThanOneReviewPerPos() {
        // given a new review (id is null), since the duplicate check runs only on creation
        Review review = TestFixtures.getReviewFixturesForInsertion().getFirst();
        Pos pos = review.pos();
        User author = review.author();
        assertNotNull(pos.getId());

        when(posDataService.getById(pos.getId())).thenReturn(pos);
        when(reviewDataService.filter(pos, author)).thenReturn(List.of(review));

        // when, then
        assertThrows(ValidationException.class, () -> reviewService.upsert(review));
        verify(posDataService).getById(pos.getId());
        verify(reviewDataService).filter(pos, author);
    }

    /**
     * Verifies that the approval status of a previously unapproved review is updated correctly
     * if it reaches the configured approval threshold.
     */
    @Test
    void approvalApprovesReviewWhenThresholdIsReached() {
        // given
        Review unapprovedReview = TestFixtures.getReviewFixtures().getFirst().toBuilder()
                .approvalCount(approvalConfiguration.minCount()-1)
                .approved(false)
                .build();

        // when
        Review updatedReview = reviewService.updateApprovalStatus(unapprovedReview);

        // then
        assertFalse(updatedReview.approved());

        // when
        Review approvedReview = unapprovedReview.toBuilder()
                .approvalCount(approvalConfiguration.minCount())
                .build();
        updatedReview = reviewService.updateApprovalStatus(approvedReview);

        // then
        assertTrue(updatedReview.approved());
    }

    /**
     * Verifies that the approval increments the count but does not mark the review as approved
     * when the approval quorum has not yet been reached.
     */
    @Test
    void approvalDoesNotApproveReviewWhenThresholdIsNotReached() {
        // given
        Review review = TestFixtures.getReviewFixtures().getFirst().toBuilder()
                .approvalCount(0)
                .approved(false)
                .build();
        assertNotNull(review.getId());
        User user = TestFixtures.getUserFixtures().getLast();
        assertNotNull(user.getId());

        when(userDataService.getById(user.getId())).thenReturn(user); // user exists
        when(reviewDataService.getById(review.getId())).thenReturn(review); // review exists
        when(reviewDataService.upsert(any(Review.class))).thenAnswer(
                invocation -> invocation.getArgument(0) // return the updated review
        );

        // when
        Review approvedReview = reviewService.approve(review.getId(), user.getId());

        // then
        verify(userDataService).getById(user.getId());
        verify(reviewDataService).getById(review.getId());
        verify(reviewDataService).upsert(any(Review.class));
        assertThat(approvedReview.approvalCount()).isEqualTo(1);
        assertThat(approvedReview.approved()).isFalse();
    }

    /**
     * Verifies that a review can be created successfully for a POS that exists
     * and for which the author has not already provided a review.
     * This covers the happy path in the review creation (previously uncovered).
     */
    @Test
    void reviewCreationSuccessfulForExistingPosAndAuthorWithoutPreviousReview() {
        // given
        Review review = TestFixtures.getReviewFixturesForInsertion().getFirst();
        Pos pos = review.pos();
        assertNotNull(pos.getId());

        when(posDataService.getById(pos.getId())).thenReturn(pos); // POS exists
        when(reviewDataService.filter(pos, review.author())).thenReturn(List.of()); // no review exists for this POS by this author
        when(reviewDataService.upsert(review)).thenReturn(review); // the data service returns the test fixture

        // when
        Review result = reviewService.upsert(review);

        // then
        verify(posDataService).getById(pos.getId());
        verify(reviewDataService).filter(pos, review.author());
        verify(reviewDataService).upsert(review);
        assertThat(result.getId()).isEqualTo(review.getId());
    }

    /**
     * Verifies that updating an existing review skips the check that runs on creation, which rejects
     * a second review by the same author for the same POS. On update that filter would match the
     * review being updated and previously made every update fail with a validation error.
     */
    @Test
    void updatingExistingReviewSkipsDuplicateCheck() {
        // given an existing review with a non-null id
        Review review = TestFixtures.getReviewFixtures().getFirst();
        Pos pos = review.pos();
        assertNotNull(review.getId());
        assertNotNull(pos.getId());

        when(posDataService.getById(pos.getId())).thenReturn(pos);
        when(reviewDataService.getById(review.getId())).thenReturn(review); // entity exists for the update
        when(reviewDataService.upsert(review)).thenReturn(review);

        // when
        Review result = reviewService.upsert(review);

        // then the per-author filter is never consulted on update
        verify(reviewDataService, never()).filter(any(Pos.class), any(User.class));
        verify(reviewDataService).upsert(review);
        assertThat(result.getId()).isEqualTo(review.getId());
    }
}

File: domain/src/test/kotlin/de/seuhd/campuscoffee/domain/implementation/ReviewServiceTest.kt
package de.seuhd.campuscoffee.domain.implementation

import de.seuhd.campuscoffee.domain.exceptions.DuplicationException
import de.seuhd.campuscoffee.domain.exceptions.NotFoundException
import de.seuhd.campuscoffee.domain.exceptions.ValidationException
import de.seuhd.campuscoffee.domain.model.objects.Review
import de.seuhd.campuscoffee.domain.ports.data.PosDataService
import de.seuhd.campuscoffee.domain.ports.data.ReviewDataService
import de.seuhd.campuscoffee.domain.ports.data.UserDataService
import de.seuhd.campuscoffee.domain.tests.TestFixtures
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

/**
 * Unit and integration tests for the operations related to reviews.
 */
@ExtendWith(MockitoExtension::class)
class ReviewServiceTest {
    private val approvalConfiguration = TestFixtures.getApprovalConfiguration()

    @Mock
    private lateinit var reviewDataService: ReviewDataService

    @Mock
    private lateinit var userDataService: UserDataService

    @Mock
    private lateinit var posDataService: PosDataService

    private lateinit var reviewService: ReviewServiceImpl

    @BeforeEach
    fun beforeEach() {
        reviewService = ReviewServiceImpl(reviewDataService, userDataService, posDataService, approvalConfiguration)
    }

    @Test
    fun `approve throws ValidationException when the user is the author`() {
        val review = TestFixtures.getReviewFixtures().first()
        val authorId = requireNotNull(review.author.id)
        whenever(userDataService.getById(authorId)).thenReturn(review.author)
        val reviewId = requireNotNull(review.id)
        whenever(reviewDataService.getById(reviewId)).thenReturn(review)

        assertThrows<ValidationException> { reviewService.approve(reviewId, authorId) }
        verify(userDataService).getById(authorId)
        verify(reviewDataService).getById(reviewId)
    }

    @Test
    fun `approve marks the review approved when the user is not the author`() {
        // one short of the quorum, so the single approval below pushes it to exactly the quorum
        val review =
            TestFixtures
                .getReviewFixtures()
                .first()
                .copy(approvalCount = approvalConfiguration.minCount - 1, approved = false)
        val user = TestFixtures.getUserFixtures().last()
        val userId = requireNotNull(user.id)
        whenever(userDataService.getById(userId)).thenReturn(user)
        val reviewId = requireNotNull(review.id)
        whenever(reviewDataService.getById(reviewId)).thenReturn(review)
        whenever(reviewDataService.upsert(any<Review>())).thenAnswer { it.getArgument<Review>(0) }

        val approvedReview = reviewService.approve(reviewId, userId)

        verify(userDataService).getById(userId)
        verify(reviewDataService).getById(reviewId)
        verify(reviewDataService).upsert(any<Review>())
        assertThat(approvedReview.approvalCount).isEqualTo(review.approvalCount + 1)
        assertThat(approvedReview.approved).isTrue()
    }

    @Test
    fun `filter returns the approved reviews for a POS`() {
        val pos = TestFixtures.getPosFixtures().first()
        val posId = requireNotNull(pos.id)
        val reviews =
            TestFixtures.getReviewFixtures().map {
                it.copy(pos = pos, approvalCount = approvalConfiguration.minCount, approved = true)
            }
        whenever(posDataService.getById(posId)).thenReturn(pos)
        whenever(reviewDataService.filter(pos, true)).thenReturn(reviews)

        val retrievedReviews = reviewService.filter(posId, true)

        verify(posDataService).getById(posId)
        verify(reviewDataService).filter(pos, true)
        assertThat(retrievedReviews).hasSize(reviews.size)
    }

    @Test
    fun `upsert throws NotFoundException when the POS does not exist`() {
        val review = TestFixtures.getReviewFixtures().first()
        val posId = requireNotNull(review.pos.id)
        whenever(posDataService.getById(posId)).thenThrow(NotFoundException(review.pos.javaClass, posId))

        assertThrows<NotFoundException> { reviewService.upsert(review) }
        verify(posDataService).getById(posId)
    }

    @Test
    fun `upsert throws DuplicationException for a duplicate review by the same author and POS`() {
        // a new review (id is null) while a persisted review by the same author for the same POS
        // exists; a duplicate is a 409 conflict, matching the uq_reviews_pos_author database constraint
        val review = TestFixtures.getReviewFixturesForInsertion().first()
        val persistedReview = TestFixtures.getReviewFixtures().first()
        val pos = review.pos
        val author = review.author
        val posId = requireNotNull(pos.id)
        whenever(posDataService.getById(posId)).thenReturn(pos)
        whenever(reviewDataService.filter(pos, author)).thenReturn(listOf(persistedReview))

        assertThrows<DuplicationException> { reviewService.upsert(review) }
        verify(posDataService).getById(posId)
        verify(reviewDataService).filter(pos, author)
    }

    @Test
    fun `updateApprovalStatus approves a review once the threshold is reached`() {
        val quorum = approvalConfiguration.minCount
        val unapprovedReview =
            TestFixtures
                .getReviewFixtures()
                .first()
                .copy(approvalCount = quorum - 1, approved = false)

        var updatedReview = reviewService.updateApprovalStatus(unapprovedReview)
        assertFalse(updatedReview.approved)

        val approvedReview = unapprovedReview.copy(approvalCount = quorum)
        updatedReview = reviewService.updateApprovalStatus(approvedReview)
        assertTrue(updatedReview.approved)
    }

    @Test
    fun `approve increments the count but leaves the review unapproved below the threshold`() {
        val review = TestFixtures.getReviewFixtures().first().copy(approvalCount = 0, approved = false)
        val reviewId = requireNotNull(review.id)
        val user = TestFixtures.getUserFixtures().last()
        val userId = requireNotNull(user.id)
        whenever(userDataService.getById(userId)).thenReturn(user)
        whenever(reviewDataService.getById(reviewId)).thenReturn(review)
        whenever(reviewDataService.upsert(any<Review>())).thenAnswer { it.getArgument<Review>(0) }

        val approvedReview = reviewService.approve(reviewId, userId)

        verify(userDataService).getById(userId)
        verify(reviewDataService).getById(reviewId)
        verify(reviewDataService).upsert(any<Review>())
        assertThat(approvedReview.approvalCount).isEqualTo(1)
        assertThat(approvedReview.approved).isFalse()
    }

    @Test
    fun `upsert saves a first review for an existing POS and author`() {
        val review = TestFixtures.getReviewFixturesForInsertion().first()
        val pos = review.pos
        val posId = requireNotNull(pos.id)
        whenever(posDataService.getById(posId)).thenReturn(pos)
        whenever(reviewDataService.filter(pos, review.author)).thenReturn(listOf())
        whenever(reviewDataService.upsert(review)).thenReturn(review)

        val result = reviewService.upsert(review)

        verify(posDataService).getById(posId)
        verify(reviewDataService).filter(pos, review.author)
        verify(reviewDataService).upsert(review)
        assertThat(result.id).isEqualTo(review.id)
    }

    @Test
    fun `upsert updates a review and keeps its approval state`() {
        // the persisted review is approved; the update carries reset values, which must not overwrite
        // the approval state managed by the approval workflow
        val existingReview = TestFixtures.getReviewFixtures().first().copy(approvalCount = 3, approved = true)
        val update =
            existingReview.copy(review = "Updated text for this review!", approvalCount = 0, approved = false)
        val pos = existingReview.pos
        val reviewId = requireNotNull(existingReview.id)
        val posId = requireNotNull(pos.id)
        whenever(posDataService.getById(posId)).thenReturn(pos)
        whenever(reviewDataService.getById(reviewId)).thenReturn(existingReview)
        whenever(reviewDataService.upsert(any<Review>())).thenAnswer { it.getArgument<Review>(0) }

        val result = reviewService.upsert(update)

        assertThat(result.review).isEqualTo(update.review)
        assertThat(result.approvalCount).isEqualTo(existingReview.approvalCount)
        assertThat(result.approved).isEqualTo(existingReview.approved)
    }

    @Test
    fun `upsert throws NotFoundException when updating a missing review`() {
        // the author already has a review for the POS, so a wrong order of checks would report the
        // duplicate conflict; the unknown id must win and yield a 404
        val persistedReview = TestFixtures.getReviewFixtures().first()
        val pos = persistedReview.pos
        val posId = requireNotNull(pos.id)
        val update = persistedReview.copy(id = 9999L)
        whenever(posDataService.getById(posId)).thenReturn(pos)
        whenever(reviewDataService.getById(9999L)).thenThrow(NotFoundException(Review::class.java, 9999L))

        assertThrows<NotFoundException> { reviewService.upsert(update) }
        verify(reviewDataService, never()).upsert(any<Review>())
    }

    @Test
    fun `upsert throws ValidationException when an update re-points a review at a different POS`() {
        // the persisted review belongs to one POS; the update payload points at another. Moving a
        // review would carry its approvals to a POS nobody approved a review for.
        val persistedReview = TestFixtures.getReviewFixtures().last()
        val targetPos = TestFixtures.getReviewFixtures().first().pos
        val update = persistedReview.copy(pos = targetPos)
        val reviewId = requireNotNull(persistedReview.id)
        whenever(posDataService.getById(requireNotNull(targetPos.id))).thenReturn(targetPos)
        whenever(reviewDataService.getById(reviewId)).thenReturn(persistedReview)

        assertThrows<ValidationException> { reviewService.upsert(update) }
        verify(reviewDataService, never()).upsert(any<Review>())
    }

    @Test
    fun `upsert throws ValidationException when an update changes the author of a review`() {
        // changing the author could retroactively mark a review as approved by its own author
        val persistedReview = TestFixtures.getReviewFixtures().first()
        val newAuthor = TestFixtures.getUserFixtures().last()
        val update = persistedReview.copy(author = newAuthor)
        val pos = persistedReview.pos
        val reviewId = requireNotNull(persistedReview.id)
        whenever(posDataService.getById(requireNotNull(pos.id))).thenReturn(pos)
        whenever(reviewDataService.getById(reviewId)).thenReturn(persistedReview)

        assertThrows<ValidationException> { reviewService.upsert(update) }
        verify(reviewDataService, never()).upsert(any<Review>())
    }
}
